google/cloud/forseti/scanner/scanners/iam_rules_scanner.py

```diff
@@ -32,6 +35,58 @@
 LOGGER = logger.get_logger(__name__)
 
 
+def _add_bucket_ancestor_bindings(policy_data):
+    """Add bucket relevant IAM policy bindings from ancestors.
+
+    Resources can inherit policy bindings from ancestors in the resource
+    manager tree. For example: a GCS bucket inherits a 'objectViewer' role
+    from a project or folder (up in the tree).
+
+    So far the IAM rules engine only checks the set of bindings directly
+    attached to a resource (direct bindings set (DBS)). We need to add
+    relevant bindings inherited from ancestors to DBS so that these are
+    also checked for violations.
+
+    If we find one more than one binding with the same role name, we need to
+    merge the members.
+
+    NOTA BENE: this function only handles buckets and bindings relevant to
+    these at present (but can and should be expanded to handle projects and
+    folders going forward).
+
+    Args:
+        policy_data (list): list of (parent resource, iam_policy resource,
+            policy bindings) tuples to find violations in.
+    """
+    storage_iam_roles = frozenset([
+        'roles/storage.admin',
+        'roles/storage.objectViewer',
+        'roles/storage.objectCreator',
+        'roles/storage.objectAdmin',
+    ])
+    bucket_data = [(r, bs) for (r, _, bs) in policy_data if r.type == 'bucket']
+    for bucket, bindings in bucket_data:
+        ancestor_bindings = [
+            bs for (r, _, bs) in policy_data
+            if r.full_name != bucket.full_name
+            and not bucket.full_name.find(r.full_name)]
+        for anbs in ancestor_bindings:
+            for anb in anbs:
+                if anb.role_name not in storage_iam_roles or anb in bindings:
+                    continue
+                # do we have a binding with the same 'role_name' already?
+                try:
+                    [same_role_binding] = [
+                        iamb for iamb in bindings
+                        if iamb.role_name == anb.role_name]
+                except ValueError:
+                    # no bindings with the same 'role_name'
+                    bindings.append(anb)
+                else:
+                    # found a binding with the same 'role_name', merge members
+                    same_role_binding.merge(anb)
```

Comment:
How does this update the `policy_data` in the outer context of this method?  I don't see where that happens.

Reply:
Remember the other review, where we discussed this once? See https://jeffknupp.com/blog/2012/11/13/is-python-callbyvalue-or-callbyreference-neither/

Reply:
In essence `same_role_binding` is a reference to an element in the `bindings` which is a reference to data in `policy_data`. These are all reference to mutable objects and the latter can thus be changed via any one of them.